### PR TITLE
Replace `if ... in spec` with `spec.satisfies` in h* and i* packages

### PR DESCRIPTION
**Pull request by aurianer**
with the structure` if "+variant" in spec:`, checks like `if "mpi" in spec:` would be valid whereas it could be `~mpi` or `+mpi`. `satisfies` adds additional checks to avoid mistakes from the user.

### CODE REVIEW

**Review by mwkrentel — APPROVED** [review]
I approve for hpctoolkit and intel-xed.

File: var/spack/repos/builtin/packages/helics/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/helics/package.py:147]: These and below appear to be breaking with conventions earlier in this file, and for consistentcy might be better as:
```suggestion
        args.append(from_variant("HELICS_DISABLE_C_SHARED_LIB", "c_shared"))
```

File: var/spack/repos/builtin/packages/ipopt/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/ipopt/package.py]: This seems incorrect and maybe should be:
```suggestion
        if spec.satisfies("^coinhsl"):
```

File: var/spack/repos/builtin/packages/ipopt/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/ipopt/package.py]: Same here.
```suggestion
        if spec.satisfies("^metis"):
```

File: var/spack/repos/builtin/packages/hypar/package.py
Comment (debog) [inline var/spack/repos/builtin/packages/hypar/package.py]: `--enable-mpi` is not necessary; adding that argument doesn't do anything.

Comment (wdconinc) [reply var/spack/repos/builtin/packages/hypar/package.py]: Good to know. This will be best addressed in a separate PR to keep this PR on scope.

**Review by wraith1995 — APPROVED** [review]
I approve for halide.

File: var/spack/repos/builtin/packages/hypar/package.py
Comment (debog) [reply var/spack/repos/builtin/packages/hypar/package.py]: Thanks; I have addressed this in a [new PR](https://github.com/spack/spack/pull/46394). I also needed to update the HyPar's repo and url since it moved to github.

**Review by debog — APPROVED** [review]

**Review by mkstoyanov — APPROVED** [review]

**Comment by mkstoyanov** [comment]
This should probably have been multiple PRs, one per-package or a few packages together. Right now, we need a dozen approvals before anything can be merged.

Anyhow, the change for heffte seems good.

**Review by alexreinking — APPROVED** [review]
Looks good for Halide

**Review by diehlpk — APPROVED** [review]
LGTM!

**Comment by diehlpk** [comment]
I approved it for hpx and hpx-kokkos.

File: var/spack/repos/builtin/packages/helics/package.py
Comment (aurianer) [reply var/spack/repos/builtin/packages/helics/package.py:147]: I guess it's opposite, with this change, it would disable c shared libs when the variant is set.

Comment (wdconinc) [reply var/spack/repos/builtin/packages/helics/package.py:147]: Right, good catch. We don't have an inversion in `define_from_variant`, so let's just leave it as is then.

File: var/spack/repos/builtin/packages/ipopt/package.py
Comment (aurianer) [reply var/spack/repos/builtin/packages/ipopt/package.py]: in this case, it's equivalent since it `depends_on(coninhsl)` when `+coinhsl`

File: var/spack/repos/builtin/packages/ipopt/package.py
Comment (aurianer) [reply var/spack/repos/builtin/packages/ipopt/package.py]: here there is a condition on the versions, so the dependency makes sense. I will change both metis and coinhsl.

**Comment by aurianer** [comment]
I had to rebase to solve a conflict

File: var/spack/repos/builtin/packages/hypar/package.py
Comment (aurianer) [reply var/spack/repos/builtin/packages/hypar/package.py]: I rebased with the latest develop and resolved the conflict :)

**Review by srekolam — APPROVED** [review]
LGTM .reviewed the hip package.

**Review by alkino — APPROVED** [review]
good for me for `highfive`, sorry for delay

**Review by wdconinc — APPROVED** [review]
I'm approving this and will leave for 24 hours to give remaining maintainers a chance to comment if they have concerns (and to give the cray runners in CI a chance to recover).

**Review by nightlark — COMMENTED** [review]
Already merged, but for the change to the helics package - LGTM

**Pull request merged**
